### PR TITLE
Add and use docker image for Emscripten builds.

### LINKED ISSUE #2963: Docker image to wrap z3 behind http
For remix, it would be helpful to have a docker image that exposes z3 / smtlib via HTTP so that it can be used from within the browser.

The idea would be to run the image either locally or on a server and connect remix to it via ajax requests.

The image should have an open port with an http server behind it, which forwards the request payload to the standard input of a z3 instance run using something like `z3 -smt2 -in` and responds with the contents of standard output of the command.

Example smtlib2 inputs look like:
```
(declare-const a Int)
(declare-fun f (Int Bool) Int)
(assert (> a 10))
(assert (< (f a true) 100))
(check-sat)
(get-model)
```

The output of z3 on this would be:
```
sat
(model 
  (define-fun a () Int
    11)
  (define-fun f ((x!1 Int) (x!2 Bool)) Int
    (ite (and (= x!1 11) (= x!2 true)) 0
      0))
)
```

During a single compilation run, multiple of these requests would be sent to the z3 "server".

### LINKED ISSUE #8851: Build and release soljson.js with SMT support.
It now seems possible to build soljson.js with integrated Z3 support.
The resulting binary seems to work, but is significantly larger, i.e. ~~30MB~~ 19MB instead of 8MB.

We can either build with Z3 by default or - if the increased size is a major concern - publish a second version (maybe marked as experimental) with Z3 enabled.

For the build process, I think it would make sense to finally also create a docker file for the emscripten build that contains prebuilt z3 and boost.

Steps used to produce a working SMT enabled soljson.js:
-------------------

I used the docker image we use for our CI runs as basis: trzeci/emscripten:sdk-tag-1.39.3-64bit

I checked out the 4.8.7 tag of Z3 and applied https://github.com/Z3Prover/z3/commit/e212159f4e941c78fc03239e0884f2f0454f581f to it.
Then I was able to build z3 with emscripten as follows:
```
emconfigure python scripts/mk_make.py --staticlib --staticbin --single-threaded --x86 --prefix=/emsdk_portable/emscripten/sdk/system/
emmake make -C build/ -j8
emmake make -C build/ install
```

Then I changed ``find_package(Z3 4.6.0)`` to ``find_package(Z3)`` in ``libsolidity/CMakeLists.txt`` (providing an up to date native /usr/bin/z3 might make this unnecessary).

Then ``scripts/travis-emscripten/install_deps.sh`` and ``scripts/travis-emscripten/build_emscripten.sh``, should build a soljson.js, but still without Z3 support.

Rerunning cmake in ``emscripten_build`` as ``cmake -DZ3_LIBRARY=/emsdk_portable/emscripten/sdk/system/lib/libz3.a -DZ3_INCLUDE_DIR=/emsdk_portable/emscripten/sdk/system/include`` and rerunning ``make`` builds ``soljson.js`` with Z3 support in ``emscripten_build/libsolc``.

TODO:
--------
- clean up this build process. We can try building using ``emcmake.py`` instead of ``emconfigure``, that way we might get cmake files that will help in ``find_package(Z3)``'s auto-detection.
- pack the Z3 build (and while we're at it the boost build) into a docker image and switch the CI builds to it
- have a look into https://github.com/cpitclaudel/cvc4.js/blob/master/vm/provision.sh and figure out if we can also enable CVC4 and what that means for binary size and if it's worth it.
- decide if the soljson.js builds should have z3 integrated by default or whether we want separate builds
- Check if any compile time options for Z3 are available that could reduce binary size.

### LINKED ISSUE #8877: Update emscripten to 1.39.15
It may be time for the "bi-annual emscripten upgrade". Last we updated to 1.39.3 it in #7832 in November.

Since then Emscripten has made 12 releases. The [interesting changes](https://github.com/emscripten-core/emscripten/blob/master/ChangeLog.md) (going from oldest release):
- Added support for streaming Wasm compilation in MINIMAL_RUNTIME (off by default)
- All ports now install their headers into a shared directory under EM_CACHE.
- Add fine-grained options for specific legacy browser support, MIN_FIREFOX_VERSION, MIN_SAFARI_VERSION, MIN_IE_VERSION, MIN_EDGE_VERSION, MIN_CHROME_VERSION. The existing LEGACY_VM_SUPPORT option sets all of them to 0, that is, maximal backwards compatibility.
- SAFE_STACK has been removed, as it overlaps with STACK_OVERFLOW_CHECK. Replace SAFE_STACK=1 with STACK_OVERFLOW_CHECK=2 (note the value is 2). This also has the effect of enabling stack checking on upstream builds when ASSERTIONS are enabled (as assertions enable STACK_OVERFLOW_CHECK=2).
- Reduced default geometric memory overgrowth rate from a very generous 2x factor to a more memory conserving +20% factor, and capped maximum reservation to 96MB at most.
- Add new Fibers API for context switching, that supersedes the old coroutine API that only ran on fastcomp. See #9859
- **Added new linker option -s WASM=2 which produces a dual Wasm+JS build, which falls back to using a JavaScript version if WebAssembly is not supported in target browser/shell. (#10118)**
- Added new linker option -s CLOSURE_WARNINGS=quiet|warn|error that allows aborting the build if the Closure compiler produced any warnings.
- wasm2js optimizations. See binaryen#2623.
- **Updated of libc++abi and libc++ to llvm 9.0.0 (#10510)** (Perhaps C++20 support?)
- The default c++ version is no longer fixed at c++03. We now fall back to clang's default which is currently c++14.
- Remove hacks from memset handling, in particular, in the wasm backend, completely remove the JS version of memset from the JS library and from DEFAULT_LIBRARY_FUNCS_TO_INCLUDE. The regular C version will be linked in from compiler_rt normally.
- Minimal runtime code size optimizations, see #10725, #10724, #10663.
- **Support for WebAssembly BigInt integration with a new WASM_BIGINT flag. With that the VM will use a JS BigInt for a wasm i64, avoiding the need for JS legalization. See #10860.** (SPEEDUP)
- Add --extern-pre-js and --extern-post-js emcc flags. Files provided there are prepended/appended to the final JavaScript output, after all other work has been done, including optimization, optional MODULARIZE-ation, instrumentation like SAFE_HEAP, etc. 
- **Update libcxx and libcxxabi to LLVM 10 release branch (#11038).** (Maybe this really brings C++20 support?)

Just for the dual wasm+js option it may be worth it to consider 1.39.7. And for BigInt 1.39.13.

cc @ekpyron

### PR DESCRIPTION
**Pull request by ekpyron**
Fixes https://github.com/ethereum/solidity/issues/8877
Fixes https://github.com/ethereum/solidity/issues/8851
Fixes https://github.com/ethereum/solidity/issues/2963

### CODE REVIEW

**Comment by ekpyron** [comment]
This is by the way ready and just waiting for @leonardoalt to adjust the tests in the solc-js repository.

**Comment by leonardoalt** [comment]
solc-js test fix: https://github.com/ethereum/solc-js/pull/466

**Review by leonardoalt — APPROVED** [review]

File: .circleci/docker/Dockerfile.emscripten [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+		-DZ3_BUILD_LIBZ3_SHARED=OFF \
+		-DZ3_ENABLE_EXAMPLE_TARGETS=OFF \
+		-DZ3_BUILD_TEST_EXECUTABLES=OFF \
+		-DZ3_BUILD_EXECUTABLE=OFF \
+		-DZ3_SINGLE_THREADED=ON \
+		..; \
+	make; make install; \
+	rm -r /usr/src/z3; \
+	cd /usr/src; \
+	wget -q 'https://dl.bintray.com/boostorg/release/1.73.0/source/boost_1_73_0.tar.bz2' -O boost.tar.bz2; \
```
Comment (axic) [inline .circleci/docker/Dockerfile.emscripten:54]: @ekpyron @chriseth So this silently also upgraded boost 1.70 (in `install_deps.sh`) to 1.73. Should include that as well as the emscripten bump in the changelog.

Comment (ekpyron) [reply .circleci/docker/Dockerfile.emscripten:54]: Yeah, true.

File: .circleci/docker/Dockerfile.emscripten [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+# (c) 2016-2019 solidity contributors.
+#------------------------------------------------------------------------------
+#
+# The Emscripten SDK at https://github.com/emscripten-core/emsdk/
+# contains a Makefile in the docker/ subdirectory that can be used to create the
+# required base image using:
+#
+#   make version=1.39.15-fastcomp build
+#
+# TODO: switch to the upstream backend by removing "-fastcomp".
```
Comment (axic) [inline .circleci/docker/Dockerfile.emscripten:31]: @ekpyron it seems we have discussed this on gitter, but what was the issue with using the new upstream backend?

Comment (ekpyron) [reply .circleci/docker/Dockerfile.emscripten:31]: It starts throwing "unknown" exceptions without apparent reason. I'm not entirely sure anymore, if it *always* does that or only for calls to Z3 - in any case, we need to figure out where those are coming from, if we want to switch.

**Pull request merged**